scripts/linters/pylint_extensions_test.py

```diff
@@ -319,21 +319,23 @@ def setUp(self):
         self.checker_test_object.setup_method()
 
     def test_no_newline_below_class_docstring(self):
-        node_no_newline_below_class_docstring = astroid.scoped_nodes.Module(
-            name='test',
-            doc='Custom test')
+        test_code = u"""
+                        class ClassName(dummy_class):
+                            \"\"\"This is a docstring.\"\"\"
+                            a = 1 + 2
+                    """
+        node_no_newline_below_class_docstring = astroid.extract_node(
+            test_code)
+
+        module_no_newline_below_class_docstring = (
+            node_no_newline_below_class_docstring.root())
```

Comment:
Doesn't this mean that you are again converting it into a module node? I think that's what using `.root()` does here. If that's true then how is it different from existing situation? (Just FYI, you can access file and filename directly also like `module_no_newline_below_class_docstring.root().file` without converting the node itself to its root)

I think we can also find a way to reach the start of docstring without using a for loop (check [here](http://pylint.pycqa.org/projects/astroid/en/latest/api/astroid.nodes.html#astroid.nodes.FunctionDef)), and then we won't need to write into a temp file and specify file, filename etc for the node, making the tests much more simpler where you only need to create a classDef node.

Also can you provide some proof of how you verified that tests weren't working properly earlier, and after you made changes they did? (Maybe print something on the terminal to demonstrate that, and show us a screenshot of it)

Reply:
No, previously we were passing `module node` to the `visit_classdef()` however, this time the `module node` is being made in order to replicate the actual scenario only and `classdef node` is only being passed to the `visit_classdef()`.

 I already checked the documentation but and I think the current method might be the only method to get the raw source code.

Reply:
> No, previously we were passing module node to the visit_classdef() however, this time the module node is being made in order to replicate the actual scenario only and classdef node is only being passed to the visit_classdef().

Ohh sorry I got confused with variable names a bit, now I see that both are separate. Thanks for clarifying!

> I already checked the documentation but and I think the current method might be the only method to get the raw source code.

Actually there is a way. Take a look at `pylint_extensions.py` file in #11450 - you can see the author has used a trick to understand where the docstring is ending using by splitting node.doc - see [here](https://github.com/oppia/oppia/pull/11450/files#diff-900df16033a01355552cd3477b8fa114f0fbb1506fb740207a7ec7a803009aacR1918).

Reply:
>Actually there is a way. Take a look at pylint_extensions.py file in #11450 - you can see the author has used a trick to understand where the docstring is ending using by splitting node.doc - see here.

@sajalasati I did thought about doing something like that but unfortunately it will not work, the author of that PR has assumed that the function's arguments will not extent more than a line.
Just check out this ss, I have replicated what he is doing...
![image](https://user-images.githubusercontent.com/52757132/113655996-930a3280-96b8-11eb-8945-db619b455bde.png)
you got my point?

Reply:
@tanishq67 The PR author did handle that case, please look at the full code (I think line 1919-1921 covers the case).

Reply:
>Also can you provide some proof of how you verified that tests weren't working properly earlier, and after you made changes they did? (Maybe print something on the terminal to demonstrate that, and show us a screenshot of it)

I added this statement to print the node.doc 
![Screenshot from 2021-04-15 10-03-03](https://user-images.githubusercontent.com/52757132/114815243-75814b00-9dd3-11eb-8c76-3bccbd01ec4d.png)
Checkout the result
![Screenshot from 2021-04-15 10-14-48](https://user-images.githubusercontent.com/52757132/114815302-8f229280-9dd3-11eb-8ca1-977740d2d45a.png)
It is printing the doc of module node - 
![Screenshot from 2021-04-15 10-04-21](https://user-images.githubusercontent.com/52757132/114815372-b11c1500-9dd3-11eb-886d-1e27d1cd964e.png)

Is this convincing ?

Reply:
> @tanishq67 The PR author did handle that case, please look at the full code (I think line 1919-1921 covers the case).

@sajalasati I think we can do that. I'll try and if it works I'll update all the ten tests, else I'll let you know. Thanks!

Reply:
> Is this convincing ?

@tanishq67 These screenshots only shows that execution reached the print statement line in your code, but how exactly do you demonstrate that earlier test wasn't working properly and the modified one does?

And how it it possible that those tests were passing earlier if they were passing the wrong node only? Go through the old code again and make sure you properly understand how it works.

Regarding what you are expected to do - you should probably first decide the logic that should be used inside `visit_classdef` function so that both the old checks work as well as the new check can be added, and then modify old tests/write new tests according to that logic. Is it clear now?

Reply:
>@tanishq67 These screenshots only shows that execution reached the print statement line in your code, but how exactly do you demonstrate that earlier test wasn't working properly and the modified one does?

@sajalasati I was trying to show that the tests are passing module node since, otherwise it should have printed this -

![image](https://user-images.githubusercontent.com/52757132/114824248-41f9ed00-9de2-11eb-9e1e-e56486ed6d91.png)

in the node.doc , dont you think so ?

Reply:
Why my test are working are legit -
They are passing `classdef` Node in the `visit_classdef()` proof - 
![Screenshot from 2021-04-15 12-12-41](https://user-images.githubusercontent.com/52757132/114825653-2263c400-9de4-11eb-89a8-2665b274bf19.png)

Reply:
>Regarding what you are expected to do - you should probably first decide the logic that should be used inside visit_classdef function so that both the old checks work as well as the new check can be added, and then modify old tests/write new tests according to that logic. Is it clear now?

The current method in `visit_classdef()` is bruteforce based **but works perfectly**, that's I am trying to fix the tests.

Reply:
@sajalasati Please let me know your thoughts. Thanks!

Reply:
>And how it it possible that those tests were passing earlier if they were passing the wrong node only? Go through the old code again and make sure you properly understand how it works.

![Screenshot from 2021-04-15 12-42-52](https://user-images.githubusercontent.com/52757132/114828980-2691e080-9de8-11eb-9032-0698a25ae8ce.png)

As the previous code was processing the raw file ( it is using `node.root().file` ) and searching for `"'`, it will reach the class's docstring and hence, unit tests will pass.
Also, an edge case which is not being taken of, is if in a test I wrote two classes in the same file, it will only check the first one. ( this will be covered if we will pass `classdef node` to `visit_classdef()`)

Reply:
@tanishq67 Thanks for explaning the difference in outputs, I get it now. Let me know once you have fixed the checker code (based on my suggestions [here](https://github.com/oppia/oppia/pull/12279#discussion_r606631926)), I will take another full pass on this PR. Thanks!

Reply:
@sajalasati There's one more edge case not handled by the author of the PR- 

![Screenshot from 2021-04-17 09-28-54](https://user-images.githubusercontent.com/52757132/115101157-7646e800-9f5f-11eb-9162-82fea225aeb4.png)

How, can we assume that it will have only one decorator ?
Moreover, any decorator can exceed 80 word limit as well, hence it will take up more than a line.

I think the current solution is handling all the edge cases (may be that's why the PR previously got merged) , Please let me know your thoughts so that I can complete the issue.

Reply:
@tanishq67 I didn't mean that the author's solution is perfect, but you can draw inspiration from it as to how to avoid reading lines from file. Just for providing better context, can you point us to examples of these in our codebase if you found any. (It's okay if you didn't find any, I am just asking)

> How, can we assume that it will have only one decorator ?

@DubeySandeep Do you think this is something @tanishq67 should consider? After looking at some files (in `controllers/`), I am not sure if we promote use of multiple decorators.

> Moreover, any decorator can exceed 80 word limit as well, hence it will take up more than a line.

@tanishq67 Can we use a logic similar to `node.args.args` or `node.doc` (whichever works, or thinking around similar lines) for incrementing the line number accordingly?

Reply:
>@tanishq67 I didn't mean that the author's solution is perfect, but you can draw inspiration from it as to how to avoid reading lines from file. Just for providing better context, can you point us to examples of these in our codebase if you found any. (It's okay if you didn't find any, I am just asking)

@sajalasati I looked at couple of different files but didn't found any case.

Reply:
>@DubeySandeep Do you think this is something @tanishq67 should consider? After looking at some files (in controllers/), I am not sure if we promote use of multiple decorators.

@DubeySandeep